kratos/utilities/normal_calculation_utils.cpp

```diff
@@ -69,99 +289,106 @@ void NormalCalculationUtils::CalculateOnSimplex(
+    this->CalculateOnSimplex(rModelPart.Conditions(), dimension_in_model_part);
+
+    // Synchronize the normal
+    rModelPart.GetCommunicator().AssembleCurrentData(NORMAL);
+}
 
-        for(auto& r_cond : rModelPart.Conditions()) {
-            for(auto& r_node: r_cond.GetGeometry()) {
-                r_node.Set(VISITED, true);
-            }
-        }
+/***********************************************************************************/
+/***********************************************************************************/
 
-        rModelPart.GetCommunicator().SynchronizeOrNodalFlags(VISITED);
+void NormalCalculationUtils::SwapNormals(ModelPart& rModelPart)
+{
+    KRATOS_TRY
 
-        for(auto& r_node: rModelPart.Nodes()) {
-            if(r_node.Is(VISITED)) {
-                r_node.FastGetSolutionStepValue(NORMAL) = zero;
-            }
-        }
-    } else {
-        // In serial iteratre normally over the condition nodes
-        for(auto& r_cond: rModelPart.Conditions()) {
-            for(auto& r_node: r_cond.GetGeometry()) {
-                r_node.FastGetSolutionStepValue(NORMAL) = zero;
-            }
-        }
+    for(auto& r_cond : rModelPart.Conditions()) {
+        GeometryType& r_geometry = r_cond.GetGeometry();
+        Node<3>::Pointer paux = r_geometry(0);
+        r_geometry(0) = r_geometry(1);
+        r_geometry(1) = paux;
     }
 
-    this->CalculateOnSimplex(rModelPart.Conditions(),Dimension);
-    rModelPart.GetCommunicator().AssembleCurrentData(NORMAL);
+    KRATOS_CATCH("")
 }
 
 /***********************************************************************************/
 /***********************************************************************************/
 
-void NormalCalculationUtils::SwapNormals(ModelPart& rModelPart)
+void NormalCalculationUtils::ComputeUnitNormalsFromAreaNormals(ModelPart& rModelPart)
 {
-    KRATOS_TRY
+    // We iterate over nodes
+    auto& r_nodes_array = rModelPart.Nodes();
+    const auto it_node_begin = r_nodes_array.begin();
+    const int num_nodes = static_cast<int>(r_nodes_array.size());
 
-    for(ModelPart::ConditionsContainerType::iterator it=rModelPart.ConditionsBegin(); it!=rModelPart.ConditionsEnd(); it++) {
-        Node<3>::Pointer paux = it->GetGeometry()(0);
-        it->GetGeometry()(0) = it->GetGeometry()(1);
-        it->GetGeometry()(1) = paux;
-        //it->GetGeometry()(0).swap(it->GetGeometry()(1));
+    #pragma omp parallel for
+    for (int i = 0; i < num_nodes; ++i) {
+        auto it_node = it_node_begin + i;
+
+        array_1d<double, 3>& r_normal = it_node->FastGetSolutionStepValue(NORMAL);
+        const double norm_normal = norm_2(r_normal);
+
+        if (norm_normal > std::numeric_limits<double>::epsilon()) r_normal /= norm_normal;
+        else KRATOS_ERROR_IF(it_node->Is(INTERFACE)) << "ERROR:: ZERO NORM NORMAL IN NODE: " << it_node->Id() << std::endl;
     }
 
-    KRATOS_CATCH("")
```

Comment:
Would it be good to have KRATOS_TRY/KRATOS_CATCH here since you are having an error thrown so the location of the error can be determined easily with full debug?

Reply:
I agree